### ISSUE TITLE
Validation and load model for evaluation is incorrect with same data as train

### ISSUE THREAD

**Issue by M6stafa** [keras-team/keras#11174]
Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

Hi,
I use keras==2.2.2 with tensorflow-gpu==1.10.0 on windows10.
As rex-yue-wu mentioned [here](https://github.com/keras-team/keras/issues/4531#issuecomment-421679734), when I train model with every other optimizer except SGD, the validation and evaluation is incorrect. I use the same data for train and validation, but the validation loss is always in a range like [1., 0.9] but the train loss goes down to 0.010.
I save the model with ModelCheckpoint callback and load and evaluate model with the following code, but get same values as validation that is for model before train:
```
def create_dense(input, units, activation_method, has_dropout):
    dense = Dense(units)(input)
    normalization = BatchNormalization()(dense)
    activation = activation_method(normalization)
    if not has_dropout:
        return activation
    else:
        dropout = Dropout(DROP_RATE)(activation)
        return dropout


def create_model(batch_size):
    model_input = Input(batch_shape=(batch_size,) + INPUT_SHAPE, dtype='float32')

    dense = create_dense(model_input, 380, LeakyReLU(alpha=LRELU_ALPHA), True)

    rnn = Reshape((1, 380))(dense)
    rnn = GRU(200, stateful=True)(rnn)

    concat = Concatenate(axis=-1)([dense, rnn])

    output = Dense(2 * NUM_OF_OUTPUTS, activation='relu')(concat)
    output = Reshape(OUTPUT_SHAPE)(output)
    output = Softmax(axis=-1)(output)

    model = Model(inputs=[model_input], outputs=[output])
    return model

def compile_model(model, learning_rate=1e-3):
    # optimizer = Adam(lr=learning_rate, beta_1=0.9, beta_2=0.999)
    optimizer = SGD(lr=learning_rate, momentum=0.9, nesterov=False)

    model.compile(loss=custom_loss,
                  metrics=[metrics.categorical_accuracy, custom_acc],
                  optimizer=optimizer,
                  loss_weights=[1.])

    return model


model = create_model(batch_size)
model.load_weights('model.h5', by_name=True)
model = compile_model(model)
model.reset_states() # because the use of stateful RNN

logs = model.evaluate_generator(
    data_generator,
    steps=len(data_generator),
    max_queue_size=1,
    workers=0,
    use_multiprocessing=False,
    verbose=1,
)

print(logs)
```

Everything work as expected when using SGD, but with every other optimizer, validations goes wrong and with SGD it takes too long to fit the model.

remove the BatchNormalization and Dropout didn't change anything.

**Comment by gabrieldemarmiesse** [comment]
Hi,
We could work on this better if you could provide a standalone reproducible example.
Some advices:

* Don't use custom data or custom paths.
* Use random arrays or even np.ones, np.zeros or custom ones. Don't load external data.
* The example should run with Keras (and deps) alone.
* Should be Python3 compatible.
* Should not be OS specific.
* The file should reproduce the bug with *high fidelity.
* Use as few layers as possible in you neural network while preserving the bug.

Link to a gist would be appreciated.
Thanks!
